File: backend/routers/process_scan.py
"""产线工序扫码：插件 / 后焊 / 三防；SMT 人工扫码（与 AOI 等效）"""
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from database import get_db
from process_scan_service import (
    STATION_LABELS,
    backfill_process_scan_model_codes,
    list_scans_for_purchase,
    scan_process,
)
from scan_station_policy import assert_scan_kind_allowed
from smt_scan_service import scan_smt_manual
from system_auth import AuthPrincipal, require_system_auth

router = APIRouter(
    prefix="/api/process-scan",
    tags=["process-scan"],
    dependencies=[Depends(require_system_auth)],
)


class ProcessScanIn(BaseModel):
    station: str = Field(..., description="plugin / post_solder / coating")
    barcode: str
    purchase_no: str = ""
    model_code: str = ""
    operator: str = ""


class SmtScanIn(BaseModel):
    barcode: str
    result: str = Field(..., description="PASS / FAIL")
    purchase_no: str = ""
    model_code: str = ""
    operator: str = ""


class PreOvenLineConfirmIn(BaseModel):
    barcode: str = Field(..., min_length=4, max_length=64)
    action: str = Field(..., description="pass / fail")
    purchase_no: str = ""
    model_code: str = ""
    remark: str = Field("", max_length=512)


@router.get("/stations")
def api_stations():
    return [{"id": k, "label": v} for k, v in STATION_LABELS.items()]


@router.post("/scan")
def api_scan(
    body: ProcessScanIn,
    db: Session = Depends(get_db),
    principal: AuthPrincipal = Depends(require_system_auth),
):
    if principal.role == "packing":
        raise HTTPException(status_code=403, detail="包装账号仅可使用包装入库扫码")
    if principal.role == "smt_scan":
        raise HTTPException(status_code=403, detail="SMT扫码账号仅可使用 SMT 人工扫码")
    assert_scan_kind_allowed(principal, body.station)
    operator = (principal.username or principal.display_name or body.operator or "").strip()
    try:
        result = scan_process(
            db,
            station=body.station,
            barcode=body.barcode,
            operator=operator,
            purchase_no=body.purchase_no,
            model_code=body.model_code,
        )
        if result.get("status") in ("ok", "already_scanned"):
            if result.get("status") == "ok":
                db.commit()
            else:
                db.rollback()
        else:
            db.rollback()
        return result
    except Exception as e:
        db.rollback()
        raise HTTPException(status_code=500, detail=str(e)) from e


@router.post("/pre-oven-aoi/line-confirm")
def api_pre_oven_line_confirm(
    body: PreOvenLineConfirmIn,
    db: Session = Depends(get_db),
    principal: AuthPrincipal = Depends(require_system_auth),
):
    """后焊产线：炉前 AOI 真不良复判（独立于 /scan 热路径）。"""
    if principal.role == "packing":
        raise HTTPException(status_code=403, detail="包装账号不可操作")
    if principal.role == "smt_scan":
        raise HTTPException(status_code=403, detail="SMT扫码账号不可操作")
    assert_scan_kind_allowed(principal, "post_solder")
    operator = (principal.username or principal.display_name or "").strip()
    try:
        from pre_oven_aoi_qc_service import line_confirm

        result = line_confirm(
            db,
            barcode=body.barcode,
            action=body.action,
            operator=operator,
            purchase_no=body.purchase_no,
            model_code=body.model_code,
            remark=body.remark or "",
        )
        db.commit()
        return result
    except ValueError as e:
        db.rollback()
        raise HTTPException(status_code=400, detail=str(e)) from e
    except Exception as e:
        db.rollback()
        raise HTTPException(status_code=500, detail=str(e)) from e


@router.post("/smt")
def api_smt_scan(
    body: SmtScanIn,
    db: Session = Depends(get_db),
    principal: AuthPrincipal = Depends(require_system_auth),
):
    if principal.role == "packing":
        raise HTTPException(status_code=403, detail="包装账号仅可使用包装入库扫码")
    assert_scan_kind_allowed(principal, "smt")
    operator = (body.operator or principal.display_name or principal.username or "").strip()
    try:
        result = scan_smt_manual(
            db,
            barcode=body.barcode,
            result=body.result,
            purchase_no=body.purchase_no,
            model_code=body.model_code,
            operator=operator,
        )
        if result.get("status") == "ok" and result.get("is_new"):
            db.commit()
        else:
            db.rollback()
        return result
    except Exception as e:
        db.rollback()
        raise HTTPException(status_code=500, detail=str(e)) from e


@router.post("/backfill-model")
def api_backfill_model(
    limit: int = Query(50000, ge=1, le=200000),
    rematch_laser: bool = Query(False, description="是否先按贴码/镭雕重挂（较慢）"),
    db: Session = Depends(get_db),
    principal: AuthPrincipal = Depends(require_system_auth),
):
    """补全历史缺机型扫码，使订单列表插件/后焊/三防数量刷新后仍可见。"""
    if principal.role in ("packing", "smt_scan"):
        raise HTTPException(status_code=403, detail="无此权限")
    try:
        result = backfill_process_scan_model_codes(db, limit=limit, rematch_laser=rematch_laser)
        db.commit()
        return {"status": "ok", **result}
    except Exception as e:
        db.rollback()
        raise HTTPException(status_code=500, detail=str(e)) from e


@router.get("/orders/{purchase_no}")
def api_list_order_scans(
    purchase_no: str,
    station: str = "",
    keyword: str = "",
    limit: int = Query(3000, ge=1, le=20000),
    db: Session = Depends(get_db),
):
    return list_scans_for_purchase(
        db,
        purchase_no,
        station=station,
        keyword=keyword,
        limit=limit,
    )

File: frontend/src/api/processScan.ts
import { apiFetch } from '@/api/http'

export type ProcessStation = 'plugin' | 'post_solder' | 'coating'

export interface ProcessScanResult {
  status: 'ok' | 'blocked' | 'already_scanned'
  message: string
  station?: string
  station_label?: string
  barcode?: string
  purchase_no?: string | null
  model_code?: string | null
  scanned_at?: string | null
  operator?: string | null
  id?: number
  aoi_result?: string
  ict_result?: string
  pre_oven_aoi_result?: string
  pre_oven_confirm_required?: boolean
  pre_oven_fail_reason?: string
  pre_oven_fail_items?: PreOvenFailItem[]
}

export interface PreOvenFailItem {
  ref: string
  part_type?: string
  part_type_zh?: string
  defect?: string
  defect_zh?: string
  label: string
}

export interface SmtScanResult {
  status: 'ok' | 'blocked' | 'already_scanned' | 'already_aoi'
  message: string
  barcode?: string
  purchase_no?: string | null
  model_code?: string | null
  scanned_at?: string | null
  operator?: string | null
  id?: number
  aoi_result?: string
  is_new?: boolean
}

export interface ProcessScanList {
  purchase_no: string
  station: string
  total: number
  items: Array<{
    id: number
    barcode: string
    station: string
    station_label: string
    model_code?: string | null
    operator?: string | null
    scanned_at?: string | null
  }>
}

export async function submitProcessScan(payload: {
  station: ProcessStation
  barcode: string
  purchase_no?: string
  model_code?: string
  operator?: string
}) {
  return apiFetch<ProcessScanResult>('/process-scan/scan', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

/** 后焊产线：炉前 AOI 真不良复判（独立于 /scan 热路径） */
export async function confirmPreOvenAoiLine(payload: {
  barcode: string
  action: 'pass' | 'fail'
  purchase_no?: string
  model_code?: string
  remark?: string
}) {
  return apiFetch<{
    status: string
    action: string
    barcode: string
    message: string
    gate_ok?: boolean
    fail_reason?: string
    already_confirmed?: boolean
  }>('/process-scan/pre-oven-aoi/line-confirm', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export async function submitSmtScan(payload: {
  barcode: string
  result: 'PASS' | 'FAIL' | string
  purchase_no?: string
  model_code?: string
  operator?: string
}) {
  return apiFetch<SmtScanResult>('/process-scan/smt', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

/** 补全历史缺机型扫码，刷新后订单列表数量可回显 */
export async function backfillProcessScanModels(limit = 50000) {
  const q = new URLSearchParams({ limit: String(limit) })
  return apiFetch<{
    status: string
    laser_scanned: number
    laser_fixed: number
    order_scanned: number
    order_filled: number
    order_skipped_multi_model: number
  }>(`/process-scan/backfill-model?${q}`, { method: 'POST' })
}

export async function fetchProcessScans(
  purchaseNo: string,
  opts: { station?: ProcessStation | ''; keyword?: string; limit?: number } = {},
) {
  const q = new URLSearchParams()
  if (opts.station) q.set('station', opts.station)
  if (opts.keyword) q.set('keyword', opts.keyword)
  if (opts.limit) q.set('limit', String(opts.limit))
  return apiFetch<ProcessScanList>(
    `/process-scan/orders/${encodeURIComponent(purchaseNo)}?${q}`,
  )
}
